src/Ast/ValuePath.php

```diff
@@ -29,17 +29,33 @@ public function __construct(AttributePath $attributePath, Filter $filter)
         $this->filter = $filter;
     }
 
+    /**
+     * @return AttributePath
+     */
+    public function getAttributePath()
+    {
+        return $this->attributePath;
+    }
+
+    /**
+     * @return Filter
+     */
+    public function getFilter()
+    {
+        return $this->filter;
+    }
+
     public function __toString()
     {
-        return sprintf('%s[%s]', $this->attributePath, $this->filter);
+        return sprintf('%s[%s]', $this->getAttributePath(), $this->getFilter());
     }
 
     public function dump()
     {
         return [
             'ValuePath' => [
-                $this->attributePath->dump(),
-                $this->filter->dump(),
+                $this->getAttributePath()->dump(),
+                $this->getFilter()->dump(),
```

Comment:
I would leave these lines as-is, no need to use the getters internally if they directly return a single property.